consensus/core/src/block_verifier.rs

```diff
@@ -1,31 +1,289 @@
-    async fn verify(&self, _b: &Block) -> Result<(), Self::Error>;
+use crate::{
+    block::{BlockAPI, SignedBlock, VerifiedBlock},
+    context::Context,
+    error::{ConsensusError, ConsensusResult},
+    transaction::TransactionVerifier,
+};
 
-    async fn verify_all(&self, _b: &[Block]) -> Result<(), Self::Error>;
+pub(crate) trait BlockVerifier: Send + Sync + 'static {
+    /// Verifies a block's metadata and transactions.
+    /// This is called before examining a block's causal history.
+    fn verify(&self, block: &SignedBlock) -> ConsensusResult<()>;
+
+    /// Verifies a block w.r.t. ancestor blocks.
+    /// This is called after a block has complete causal history locally,
+    /// and is ready to be accepted into the DAG.
+    ///
+    /// Caller must make sure ancestors corresponse to block.ancestors() 1-to-1, in the same order.
+    fn check_ancestors(
+        &self,
+        block: &VerifiedBlock,
+        ancestors: &[VerifiedBlock],
+    ) -> ConsensusResult<()>;
+}
+
+/// `SignedBlockVerifier` checks the validity of a block.
+///
+/// Blocks that fail verification at one honest authority will be rejected by all other honest
+/// authorities as well. The means invalid blocks, and blocks with an invalid ancestor, will never
+/// be accepted into the DAG.
+pub(crate) struct SignedBlockVerifier {
+    context: Arc<Context>,
+    transaction_verifier: Arc<dyn TransactionVerifier>,
+}
+
+impl SignedBlockVerifier {
+    pub(crate) fn new(
+        context: Arc<Context>,
+        transaction_verifier: Arc<dyn TransactionVerifier>,
+    ) -> Self {
+        Self {
+            context,
+            transaction_verifier,
+        }
+    }
 }
 
-#[derive(Clone)]
-pub struct TestBlockVerifier;
+// All block verification logic are implemented below.
+impl BlockVerifier for SignedBlockVerifier {
+    fn verify(&self, block: &SignedBlock) -> ConsensusResult<()> {
+        let committee = &self.context.committee;
+        // The block must belong to the current epoch and have valid authority index,
+        // before having its signature verified.
+        if block.epoch() != committee.epoch() {
+            return Err(ConsensusError::WrongEpoch {
+                expected: committee.epoch(),
+                actual: block.epoch(),
+            });
+        }
+        if !committee.is_valid_index(block.author()) {
+            return Err(ConsensusError::InvalidAuthorityIndex {
+                index: block.author(),
+                max: committee.size() - 1,
+            });
+        }
+        // Verifiy the block's signature.
+        block.verify_signature(&self.context)?;
+        // Verify the block's ancestor refs are consistent with the block's round,
+        // and total parent stakes reach quorum.
+        if block.ancestors().len() > committee.size() {
+            return Err(ConsensusError::TooManyAncestors(
+                block.ancestors().len(),
+                committee.size(),
+            ));
+        }
```

Comment:
Yeah I think if ancestors are selected from different authorities when proposing, this would be automatically satisfied. I can take a look at Core too.